Looser type annotations for wrap_file() and FileWrapper

`werkzeug.wsgi.wrap_file()` and `FileWrapper` require the `file` argument to have type `BinaryIO`. Generally, the use of `IO` classes is discouraged, especially as arguments, since they are normally much too strict and incompatible with each other. In my particular use case, I got an error message 'Argument 2 to "wrap_file" has incompatible type "IO[bytes]"; expected "BinaryIO"'.

As `FileWrapper` only requires the `read()` method to be present (although it uses more methods if they exist), the old typeshed stubs annotated the `file` arguments with `SupportsRead[bytes]`, which is available in `_typeshed` during type checking:

```python
if t.TYPE_CHECKING:
    from _typeshed import SupportsRead
```

Alternatively a similar protocol can be defined like this:

```python
_T_co = t.TypeVar("_T_co", covariant=True)
class SupportsRead(t.Protocol[_T_co]):
    def read(self, __length: int = ...) -> _T_co: ...
```

I'm running in the same typing issue with `FileStorage`, I've extracted this snippet from our project:

```py
from zipfile import ZipFile
from werkzeug.datastructures import FileStorage


with ZipFile(FileStorage()) as zf:
    pass
```
```
test.py:5: error: Argument 1 to "ZipFile" has incompatible type "FileStorage"; expected "Union[Union[str, _PathLike[str]], IO[bytes]]"
Found 1 error in 1 file (checked 1 source file)
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -28,6 +28,9 @@ Unreleased
 -   The debugger pin cookie is set with ``SameSite`` set to ``Strict``
     instead of ``None`` to be compatible with modern browser security.
     :issue:`2156`
+-   Type annotations use ``IO[bytes]`` and ``IO[str]`` instead of
+    ``BinaryIO`` and ``TextIO`` for wider type compatibility.
+    :issue:`2130`
 
 
 Version 2.0.1

--- a/src/werkzeug/datastructures.pyi
+++ b/src/werkzeug/datastructures.pyi
@@ -1,13 +1,13 @@
 from datetime import datetime
 from os import PathLike
 from typing import Any
-from typing import BinaryIO
 from typing import Callable
 from typing import Collection
 from typing import Dict
 from typing import FrozenSet
 from typing import Generic
 from typing import Hashable
+from typing import IO
 from typing import Iterable
 from typing import Iterator
 from typing import List
@@ -344,7 +344,7 @@ class FileMultiDict(MultiDict[str, "FileStorage"]):
     def add_file(
         self,
         name: str,
-        file: Union[FileStorage, str, BinaryIO],
+        file: Union[FileStorage, str, IO[bytes]],
         filename: Optional[str] = None,
         content_type: Optional[str] = None,
     ) -> None: ...
@@ -878,13 +878,13 @@ class WWWAuthenticate(UpdateDictMixin[str, str], Dict[str, str]):
 
 class FileStorage:
     name: Optional[str]
-    stream: BinaryIO
+    stream: IO[bytes]
     filename: Optional[str]
     headers: Headers
     _parsed_content_type: Tuple[str, Dict[str, str]]
     def __init__(
         self,
-        stream: Optional[BinaryIO] = None,
+        stream: Optional[IO[bytes]] = None,
         filename: Optional[str] = None,
         name: Optional[str] = None,
         content_type: Optional[str] = None,
@@ -901,7 +901,7 @@ class FileStorage:
     @property
     def mimetype_params(self) -> Dict[str, str]: ...
     def save(
-        self, dst: Union[str, PathLike, BinaryIO], buffer_size: int = ...
+        self, dst: Union[str, PathLike, IO[bytes]], buffer_size: int = ...
     ) -> None: ...
     def close(self) -> None: ...
     def __bool__(self) -> bool: ...

--- a/src/werkzeug/debug/tbtools.py
+++ b/src/werkzeug/debug/tbtools.py
@@ -259,7 +259,7 @@ class Traceback:
         """String representation of the final exception."""
         return self.groups[-1].exception
 
-    def log(self, logfile: t.Optional[t.TextIO] = None) -> None:
+    def log(self, logfile: t.Optional[t.IO[str]] = None) -> None:
         """Log the ASCII traceback into a file object."""
         if logfile is None:
             logfile = sys.stderr

--- a/src/werkzeug/formparser.py
+++ b/src/werkzeug/formparser.py
@@ -35,7 +35,7 @@ if t.TYPE_CHECKING:
     import typing as te
     from _typeshed.wsgi import WSGIEnvironment
 
-    t_parse_result = t.Tuple[t.BinaryIO, MultiDict, MultiDict]
+    t_parse_result = t.Tuple[t.IO[bytes], MultiDict, MultiDict]
 
     class TStreamFactory(te.Protocol):
         def __call__(
@@ -44,14 +44,14 @@ if t.TYPE_CHECKING:
             content_type: t.Optional[str],
             filename: t.Optional[str],
             content_length: t.Optional[int] = None,
-        ) -> t.BinaryIO:
+        ) -> t.IO[bytes]:
             ...
 
 
 F = t.TypeVar("F", bound=t.Callable[..., t.Any])
 
 
-def _exhaust(stream: t.BinaryIO) -> None:
+def _exhaust(stream: t.IO[bytes]) -> None:
     bts = stream.read(64 * 1024)
     while bts:
         bts = stream.read(64 * 1024)
@@ -62,13 +62,13 @@ def default_stream_factory(
     content_type: t.Optional[str],
     filename: t.Optional[str],
     content_length: t.Optional[int] = None,
-) -> t.BinaryIO:
+) -> t.IO[bytes]:
     max_size = 1024 * 500
 
     if SpooledTemporaryFile is not None:
-        return t.cast(t.BinaryIO, SpooledTemporaryFile(max_size=max_size, mode="rb+"))
+        return t.cast(t.IO[bytes], SpooledTemporaryFile(max_size=max_size, mode="rb+"))
     elif total_content_length is None or total_content_length > max_size:
-        return t.cast(t.BinaryIO, TemporaryFile("rb+"))
+        return t.cast(t.IO[bytes], TemporaryFile("rb+"))
 
     return BytesIO()
 
@@ -212,7 +212,7 @@ class FormDataParser:
         self, mimetype: str, options: t.Dict[str, str]
     ) -> t.Optional[
         t.Callable[
-            ["FormDataParser", t.BinaryIO, str, t.Optional[int], t.Dict[str, str]],
+            ["FormDataParser", t.IO[bytes], str, t.Optional[int], t.Dict[str, str]],
             "t_parse_result",
         ]
     ]:
@@ -231,7 +231,7 @@ class FormDataParser:
 
     def parse(
         self,
-        stream: t.BinaryIO,
+        stream: t.IO[bytes],
         mimetype: str,
         content_length: t.Optional[int],
         options: t.Optional[t.Dict[str, str]] = None,
@@ -272,7 +272,7 @@ class FormDataParser:
     @exhaust_stream
     def _parse_multipart(
         self,
-        stream: t.BinaryIO,
+        stream: t.IO[bytes],
         mimetype: str,
         content_length: t.Optional[int],
         options: t.Dict[str, str],
@@ -295,7 +295,7 @@ class FormDataParser:
     @exhaust_stream
     def _parse_urlencoded(
         self,
-        stream: t.BinaryIO,
+        stream: t.IO[bytes],
         mimetype: str,
         content_length: t.Optional[int],
         options: t.Dict[str, str],
@@ -316,7 +316,7 @@ class FormDataParser:
     parse_functions: t.Dict[
         str,
         t.Callable[
-            ["FormDataParser", t.BinaryIO, str, t.Optional[int], t.Dict[str, str]],
+            ["FormDataParser", t.IO[bytes], str, t.Optional[int], t.Dict[str, str]],
             "t_parse_result",
         ],
     ] = {
@@ -418,7 +418,7 @@ class MultiPartParser:
 
     def start_file_streaming(
         self, event: File, total_content_length: t.Optional[int]
-    ) -> t.BinaryIO:
+    ) -> t.IO[bytes]:
         content_type = event.headers.get("content-type")
 
         try:
@@ -435,9 +435,9 @@ class MultiPartParser:
         return container
 
     def parse(
-        self, stream: t.BinaryIO, boundary: bytes, content_length: t.Optional[int]
+        self, stream: t.IO[bytes], boundary: bytes, content_length: t.Optional[int]
     ) -> t.Tuple[MultiDict, MultiDict]:
-        container: t.Union[t.BinaryIO, t.List[bytes]]
+        container: t.Union[t.IO[bytes], t.List[bytes]]
         _write: t.Callable[[bytes], t.Any]
 
         iterator = chain(
@@ -476,7 +476,7 @@ class MultiPartParser:
                             )
                             fields.append((current_part.name, value))
                         else:
-                            container = t.cast(t.BinaryIO, container)
+                            container = t.cast(t.IO[bytes], container)
                             container.seek(0)
                             files.append(
                                 (

--- a/src/werkzeug/middleware/lint.py
+++ b/src/werkzeug/middleware/lint.py
@@ -45,7 +45,7 @@ def check_type(context: str, obj: object, need: t.Type = str) -> None:
 
 
 class InputStream:
-    def __init__(self, stream: t.BinaryIO) -> None:
+    def __init__(self, stream: t.IO[bytes]) -> None:
         self._stream = stream
 
     def read(self, *args: t.Any) -> bytes:
@@ -97,7 +97,7 @@ class InputStream:
 
 
 class ErrorStream:
-    def __init__(self, stream: t.TextIO) -> None:
+    def __init__(self, stream: t.IO[str]) -> None:
         self._stream = stream
 
     def write(self, s: str) -> None:

--- a/src/werkzeug/middleware/profiler.py
+++ b/src/werkzeug/middleware/profiler.py
@@ -77,7 +77,7 @@ class ProfilerMiddleware:
     def __init__(
         self,
         app: "WSGIApplication",
-        stream: t.TextIO = sys.stdout,
+        stream: t.IO[str] = sys.stdout,
         sort_by: t.Iterable[str] = ("time", "calls"),
         restrictions: t.Iterable[t.Union[str, int, float]] = (),
         profile_dir: t.Optional[str] = None,

--- a/src/werkzeug/middleware/shared_data.py
+++ b/src/werkzeug/middleware/shared_data.py
@@ -27,7 +27,7 @@ from ..utils import get_content_type
 from ..wsgi import get_path_info
 from ..wsgi import wrap_file
 
-_TOpener = t.Callable[[], t.Tuple[t.BinaryIO, datetime, int]]
+_TOpener = t.Callable[[], t.Tuple[t.IO[bytes], datetime, int]]
 _TLoader = t.Callable[[t.Optional[str]], t.Tuple[t.Optional[str], t.Optional[_TOpener]]]
 
 if t.TYPE_CHECKING:

--- a/src/werkzeug/serving.py
+++ b/src/werkzeug/serving.py
@@ -86,7 +86,7 @@ if t.TYPE_CHECKING:
 class DechunkedInput(io.RawIOBase):
     """An input stream that handles Transfer-Encoding 'chunked'"""
 
-    def __init__(self, rfile: t.BinaryIO) -> None:
+    def __init__(self, rfile: t.IO[bytes]) -> None:
         self._rfile = rfile
         self._done = False
         self._len = 0

--- a/src/werkzeug/test.py
+++ b/src/werkzeug/test.py
@@ -56,7 +56,7 @@ def stream_encode_multipart(
     threshold: int = 1024 * 500,
     boundary: t.Optional[str] = None,
     charset: str = "utf-8",
-) -> t.Tuple[t.BinaryIO, int, str]:
+) -> t.Tuple[t.IO[bytes], int, str]:
     """Encode a dict of values (either strings or file descriptors or
     :class:`FileStorage` objects.) into a multipart encoded string stored
     in a file descriptor.
@@ -64,7 +64,7 @@ def stream_encode_multipart(
     if boundary is None:
         boundary = f"---------------WerkzeugFormPart_{time()}{random()}"
 
-    stream: t.BinaryIO = BytesIO()
+    stream: t.IO[bytes] = BytesIO()
     total_length = 0
     on_disk = False
 
@@ -81,7 +81,7 @@ def stream_encode_multipart(
                 if length + total_length <= threshold:
                     stream.write(s)
                 else:
-                    new_stream = t.cast(t.BinaryIO, TemporaryFile("wb+"))
+                    new_stream = t.cast(t.IO[bytes], TemporaryFile("wb+"))
                     new_stream.write(stream.getvalue())  # type: ignore
                     new_stream.write(s)
                     stream = new_stream
@@ -347,7 +347,7 @@ class EnvironBuilder:
 
     _args: t.Optional[MultiDict]
     _query_string: t.Optional[str]
-    _input_stream: t.Optional[t.BinaryIO]
+    _input_stream: t.Optional[t.IO[bytes]]
     _form: t.Optional[MultiDict]
     _files: t.Optional[FileMultiDict]
 
@@ -357,15 +357,17 @@ class EnvironBuilder:
         base_url: t.Optional[str] = None,
         query_string: t.Optional[t.Union[t.Mapping[str, str], str]] = None,
         method: str = "GET",
-        input_stream: t.Optional[t.BinaryIO] = None,
+        input_stream: t.Optional[t.IO[bytes]] = None,
         content_type: t.Optional[str] = None,
         content_length: t.Optional[int] = None,
-        errors_stream: t.Optional[t.TextIO] = None,
+        errors_stream: t.Optional[t.IO[str]] = None,
         multithread: bool = False,
         multiprocess: bool = False,
         run_once: bool = False,
         headers: t.Optional[t.Union[Headers, t.Iterable[t.Tuple[str, str]]]] = None,
-        data: t.Optional[t.Union[t.BinaryIO, str, bytes, t.Mapping[str, t.Any]]] = None,
+        data: t.Optional[
+            t.Union[t.IO[bytes], str, bytes, t.Mapping[str, t.Any]]
+        ] = None,
         environ_base: t.Optional[t.Mapping[str, t.Any]] = None,
         environ_overrides: t.Optional[t.Mapping[str, t.Any]] = None,
         charset: str = "utf-8",
@@ -490,7 +492,7 @@ class EnvironBuilder:
         self,
         key: str,
         value: t.Union[
-            t.BinaryIO, t.Tuple[t.BinaryIO, str], t.Tuple[t.BinaryIO, str, str]
+            t.IO[bytes], t.Tuple[t.IO[bytes], str], t.Tuple[t.IO[bytes], str, str]
         ],
     ) -> None:
         """Called in the EnvironBuilder to add files from the data dict."""
@@ -638,7 +640,7 @@ class EnvironBuilder:
         self._set_form("_files", value)
 
     @property
-    def input_stream(self) -> t.Optional[t.BinaryIO]:
+    def input_stream(self) -> t.Optional[t.IO[bytes]]:
         """An optional input stream. This is mutually exclusive with
         setting :attr:`form` and :attr:`files`, setting it will clear
         those. Do not provide this if the method is not ``POST`` or
@@ -647,7 +649,7 @@ class EnvironBuilder:
         return self._input_stream
 
     @input_stream.setter
-    def input_stream(self, value: t.Optional[t.BinaryIO]) -> None:
+    def input_stream(self, value: t.Optional[t.IO[bytes]]) -> None:
         self._input_stream = value
         self._form = None
         self._files = None

--- a/src/werkzeug/urls.py
+++ b/src/werkzeug/urls.py
@@ -869,7 +869,7 @@ def url_decode(
 
 
 def url_decode_stream(
-    stream: t.BinaryIO,
+    stream: t.IO[bytes],
     charset: str = "utf-8",
     decode_keys: None = None,
     include_empty: bool = True,
@@ -990,7 +990,7 @@ def url_encode(
 
 def url_encode_stream(
     obj: t.Union[t.Mapping[str, str], t.Iterable[t.Tuple[str, str]]],
-    stream: t.Optional[t.TextIO] = None,
+    stream: t.Optional[t.IO[str]] = None,
     charset: str = "utf-8",
     encode_keys: None = None,
     sort: bool = False,

--- a/src/werkzeug/utils.py
+++ b/src/werkzeug/utils.py
@@ -588,7 +588,7 @@ def append_slash_redirect(environ: "WSGIEnvironment", code: int = 301) -> "Respo
 
 
 def send_file(
-    path_or_file: t.Union[os.PathLike, str, t.BinaryIO],
+    path_or_file: t.Union[os.PathLike, str, t.IO[bytes]],
     environ: "WSGIEnvironment",
     mimetype: t.Optional[str] = None,
     as_attachment: bool = False,
@@ -681,7 +681,7 @@ def send_file(
         response_class = Response
 
     path: t.Optional[str] = None
-    file: t.Optional[t.BinaryIO] = None
+    file: t.Optional[t.IO[bytes]] = None
     size: t.Optional[int] = None
     mtime: t.Optional[float] = None
     headers = Headers()

--- a/src/werkzeug/wrappers/request.py
+++ b/src/werkzeug/wrappers/request.py
@@ -216,7 +216,7 @@ class Request(_SansIORequest):
         content_type: t.Optional[str],
         filename: t.Optional[str] = None,
         content_length: t.Optional[int] = None,
-    ) -> t.BinaryIO:
+    ) -> t.IO[bytes]:
         """Called to get a stream for the file upload.
 
         This must provide a file-like class with `read()`, `readline()`
@@ -300,7 +300,7 @@ class Request(_SansIORequest):
         d = self.__dict__
         d["stream"], d["form"], d["files"] = data
 
-    def _get_stream_for_parsing(self) -> t.BinaryIO:
+    def _get_stream_for_parsing(self) -> t.IO[bytes]:
         """This is the same as accessing :attr:`stream` with the difference
         that if it finds cached data from calling :meth:`get_data` first it
         will create a new stream out of the cached data.
@@ -330,7 +330,7 @@ class Request(_SansIORequest):
         self.close()
 
     @cached_property
-    def stream(self) -> t.BinaryIO:
+    def stream(self) -> t.IO[bytes]:
         """
         If the incoming form data was not encoded with a known mimetype
         the data is stored unmodified in this stream for consumption.  Most
@@ -355,7 +355,7 @@ class Request(_SansIORequest):
 
         return get_input_stream(self.environ)
 
-    input_stream = environ_property[t.BinaryIO](
+    input_stream = environ_property[t.IO[bytes]](
         "wsgi.input",
         doc="""The WSGI input stream.
 

--- a/src/werkzeug/wsgi.py
+++ b/src/werkzeug/wsgi.py
@@ -140,7 +140,7 @@ def get_content_length(environ: "WSGIEnvironment") -> t.Optional[int]:
 
 def get_input_stream(
     environ: "WSGIEnvironment", safe_fallback: bool = True
-) -> t.BinaryIO:
+) -> t.IO[bytes]:
     """Returns the input stream from the WSGI environment and wraps it
     in the most sensible way possible. The stream returned is not the
     raw WSGI stream in most cases but one that is safe to read from
@@ -157,7 +157,7 @@ def get_input_stream(
         content length is not set. Disabling this allows infinite streams,
         which can be a denial-of-service risk.
     """
-    stream = t.cast(t.BinaryIO, environ["wsgi.input"])
+    stream = t.cast(t.IO[bytes], environ["wsgi.input"])
     content_length = get_content_length(environ)
 
     # A wsgi extension that tells us if the input is terminated.  In
@@ -173,7 +173,7 @@ def get_input_stream(
         return io.BytesIO() if safe_fallback else stream
 
     # Otherwise limit the stream to the content length
-    return t.cast(t.BinaryIO, LimitedStream(stream, content_length))
+    return t.cast(t.IO[bytes], LimitedStream(stream, content_length))
 
 
 def get_query_string(environ: "WSGIEnvironment") -> str:
@@ -467,7 +467,7 @@ class ClosingIterator:
 
 
 def wrap_file(
-    environ: "WSGIEnvironment", file: t.BinaryIO, buffer_size: int = 8192
+    environ: "WSGIEnvironment", file: t.IO[bytes], buffer_size: int = 8192
 ) -> t.Iterable[bytes]:
     """Wraps a file.  This uses the WSGI server's file wrapper if available
     or otherwise the generic :class:`FileWrapper`.
@@ -507,7 +507,7 @@ class FileWrapper:
     :param buffer_size: number of bytes for one iteration.
     """
 
-    def __init__(self, file: t.BinaryIO, buffer_size: int = 8192) -> None:
+    def __init__(self, file: t.IO[bytes], buffer_size: int = 8192) -> None:
         self.file = file
         self.buffer_size = buffer_size
 
@@ -560,7 +560,7 @@ class _RangeWrapper:
 
     def __init__(
         self,
-        iterable: t.Union[t.Iterable[bytes], t.BinaryIO],
+        iterable: t.Union[t.Iterable[bytes], t.IO[bytes]],
         start_byte: int = 0,
         byte_range: t.Optional[int] = None,
     ):
@@ -631,7 +631,7 @@ class _RangeWrapper:
 
 
 def _make_chunk_iter(
-    stream: t.Union[t.Iterable[bytes], t.BinaryIO],
+    stream: t.Union[t.Iterable[bytes], t.IO[bytes]],
     limit: t.Optional[int],
     buffer_size: int,
 ) -> t.Iterator[bytes]:
@@ -645,9 +645,9 @@ def _make_chunk_iter(
             if item:
                 yield item
         return
-    stream = t.cast(t.BinaryIO, stream)
+    stream = t.cast(t.IO[bytes], stream)
     if not isinstance(stream, LimitedStream) and limit is not None:
-        stream = t.cast(t.BinaryIO, LimitedStream(stream, limit))
+        stream = t.cast(t.IO[bytes], LimitedStream(stream, limit))
     _read = stream.read
     while True:
         item = _read(buffer_size)
@@ -657,7 +657,7 @@ def _make_chunk_iter(
 
 
 def make_line_iter(
-    stream: t.Union[t.Iterable[bytes], t.BinaryIO],
+    stream: t.Union[t.Iterable[bytes], t.IO[bytes]],
     limit: t.Optional[int] = None,
     buffer_size: int = 10 * 1024,
     cap_at_buffer: bool = False,
@@ -749,7 +749,7 @@ def make_line_iter(
 
 
 def make_chunk_iter(
-    stream: t.Union[t.Iterable[bytes], t.BinaryIO],
+    stream: t.Union[t.Iterable[bytes], t.IO[bytes]],
     separator: bytes,
     limit: t.Optional[int] = None,
     buffer_size: int = 10 * 1024,
@@ -859,7 +859,7 @@ class LimitedStream(io.IOBase):
                   end with `EOF` (like `wsgi.input`)
     """
 
-    def __init__(self, stream: t.BinaryIO, limit: int) -> None:
+    def __init__(self, stream: t.IO[bytes], limit: int) -> None:
         self._read = stream.read
         self._readline = stream.readline
         self._pos = 0

